app/history.py

```diff
@@ -19,44 +19,81 @@
 import app.log
 import cPickle as pickle
 import os
+import hashlib
+import time
+import app.prefs
 
-data = {}
-path = None
-
-def get(keyPath, default=None):
-  global data
-  cursor = data
-  for i in keyPath[:-1]:
-    cursor = cursor.setdefault(i, {})
-  return cursor.get(keyPath[-1], default)
-
-def set(keyPath, value):
-  global data
-  cursor = data
-  for i in keyPath[:-1]:
-    cursor = cursor.setdefault(i, {})
-  cursor[keyPath[-1]] = value
-  #assert get(keyPath) == value
-
-def loadUserHistory(filePath):
-  global data, path
-  try:
-    path = filePath
-    if os.path.isfile(path):
-      with open(path, "rb") as file:
-        data = pickle.load(file)
-    else:
-      data = {}
-  except Exception, e:
-    app.log.error('exception')
-    data = {}
+userHistory = {}
+pathToHistory = app.prefs.prefs['userData'].get('historyPath')
+
+def loadUserHistory(filePath, historyPath=pathToHistory):
+  global userHistory, pathToHistory
+  pathToHistory = historyPath
+  if os.path.isfile(historyPath):
+    with open(historyPath, 'rb') as file:
+      userHistory = pickle.load(file)
 
-def saveUserHistory():
-  global data, path
+def saveUserHistory(fileInfo, fileHistory, historyPath=pathToHistory):
+  """
+  Args:
+    fileInfo (tuple): Contains (filePath, lastChecksum, lastFileSize).
+    fileHistory (dict): The history of the file that the user wants to save.
+    historyPath (str): The path to the user's saved history.
+
+  Returns:
+    None
+  """
+  global userHistory, pathToHistory
+  filePath, lastChecksum, lastFileSize = fileInfo
   try:
-    if path is not None:
-      with open(path, "wb") as file:
-        pickle.dump(data, file)
+    if historyPath is not None:
+      pathToHistory = historyPath
+      userHistory.pop((lastChecksum, lastFileSize), None)
+      newChecksum, newFileSize = getFileInfo(filePath)
+      userHistory[(newChecksum, newFileSize)] = fileHistory
+      with open(historyPath, 'wb') as file:
+        pickle.dump(userHistory, file)
       app.log.info('wrote pickle')
   except Exception, e:
     app.log.error('exception')
+
+def getFileHistory(filePath):
+  checksum, fileSize = getFileInfo(filePath)
+  fileHistory = userHistory.get((checksum, fileSize), {})
+  fileHistory['adate'] = time.time()
+  return fileHistory
+
+def getFileInfo(filePath):
+  """
+  Args:
+    filePath (str): The absolute path to the file.
+
+  Returns:
+    A tuple containing the checksum and size of the file.
+  """
+  try:
+    checksum = calculateChecksum(filePath)
+    fileSize = os.stat(filePath).st_size
+    return (checksum, fileSize)
+  except:
+    return (None, 0)
+
+def calculateChecksum(filePath):
+  app.log.info("Calculate checksum of the current file")
+  hasher = hashlib.sha512()
+  with open(filePath, 'rb') as file:
+    hasher.update(file.read())
```

Comment:
Ahh I guess I reverted the change. I wrote this part up because it would allow you to read the file in chunks so you wouldn't have to load the whole thing into memory which would cause an issue if the file was too big. However, I guess that won't really be an issue since this is just text files. I'll change it for now and assume it can read the entire file until we actually hit an issue.